README
# pretty-money

A tiny money formatting library for JS

- **Small.** The code is dead simple and the size is controlled by [Size Limit](https://github.com/ai/size-limit)
- **FP-friendly.** The functions can be curried to create the formatting function that you need.
- **Tree-shakable.** You don't have to import the whole library to only format one currency.
- **International.** It supports the nationally-accepted currency signs and their placements for every currency.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.1] - 2021-04-08

### Fixed

- invalid formatting of numbers smaller than 1 (#8)

## [1.1.0] - 2021-03-20

### Added

- this changelog
- CommonJS version separate from UMD (`pretty-money.js`)
- ESM version for use in modern browsers (`pretty-money.modern.js`)
- source maps for generated files

### Changed

- source file language from TS to JS
- files for Node (`.esm.js`, `.js`) are built unminified, files for Web (`.modern.js`, `.umd.js`) are built minified

### Security

- bumped dev dependencies

## [1.0.5] - 2020-04-04

### Changed

- NPM → Yarn for dependency management

### Security

- bumped dev dependencies

## [1.0.4] - 2020-01-17

This version improves the build process and updates the README. The package contents haven't changed.

## [1.0.3] - 2019-10-24

Version 1.0.2 was not published properly

### Changed

- TypeScript definitions to use a better, more readable overloading syntax
- bundle size reduced to 468 bytes (16% less than before)

## [1.0.1] - 2019-09-08

### Added

- bundles now have the header comment with `@license`

### Fixed

- `minDecimal` had no effect if the number had no decimal part (#2)

### Regress

- bundle size is now 19% bigger (469 → 559 bytes minzipped)

## [1.0.0] - 2019-09-08

The initial release of pretty-money ✨

### Added

- setting for currency sign and its position
- spacing setting
- setting for delimiters
- setting for decimal part formatting

[Unreleased]: https://github.com/kytta/pretty-money/compare/v1.1.1...HEAD
[1.1.1]: https://github.com/kytta/pretty-money/compare/v1.1.0...v1.1.1
[1.1.0]: https://github.com/kytta/pretty-money/compare/v1.0.5...v1.1.0
[1.0.5]: https://github.com/kytta/pretty-money/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/kytta/pretty-money/compare/1.0.3...v1.0.4
[1.0.3]: https://github.com/kytta/pretty-money/compare/1.0.1...1.0.3
[1.0.1]: https://github.com/kytta/pretty-money/compare/1.0.0...1.0.1
[1.0.0]: https://github.com/kytta/pretty-money/compare/d90fe8630c3a595c435c7b78881c6e57e51a221d...1.0.0

test/core.test.ts
import prettyMoney from "../src/pretty-money";

test("curried should match simple", () => {
    expect(prettyMoney({})(1234))
        .toBe(prettyMoney({}, 1234));
});


test/number.test.ts
import prettyMoney from "../src/pretty-money";

test("integer number", () => {
    expect(prettyMoney({}, 1234))
        .toBe("1234");
});

test("float number", () => {
    expect(prettyMoney({}, 1234.56))
        .toBe("1234.56");
});

test("integer string", () => {
    expect(prettyMoney({}, "1234"))
        .toBe("1234");
});

test("float string", () => {
    expect(prettyMoney({}, "1234.56"))
        .toBe("1234.56");
});

test("wrong string", () => {
    expect(prettyMoney({}, "1234,56"))
        .toBe("NaN");
});


test/currency.test.ts
import prettyMoney from "../src/pretty-money";

test("default config with currency", () => {
    expect(prettyMoney({ currency: "RUB" }, 1234))
        .toBe("1234 RUB");
});

test("default config with currency before amount", () => {
    expect(prettyMoney({ currency: "USD", position: "before" }, 1234))
        .toBe("USD 1234");
});


.eslintrc.js
module.exports = {
    parser: "@typescript-eslint/parser",
    extends: ["plugin:@typescript-eslint/recommended"],
    plugins: ["@typescript-eslint"],
    rules: {
        "array-bracket-spacing": 2,
        "brace-style": [2, "1tbs"],
        "camelcase": 2,
        "comma-spacing": 2,
        "comma-style": 2,
        "computed-property-spacing": 2,
        "consistent-return": 2,
        "curly": 2,
        "dot-location": [2, "property"],
        "eqeqeq": [2, "smart"],
        "func-call-spacing": 2,
        "func-names": [2, "as-needed"],
        "guard-for-in": 1,
        "key-spacing": 2,
        "keyword-spacing": 2,
        "linebreak-style": [1, "unix"],
        "no-console": 0,
        "no-delete-var": 2,
        "no-else-return": 2,
        "no-empty": 2,
        "no-mixed-operators": 2,
        "no-multi-spaces": 2,
        "no-trailing-spaces": 0,
        "no-var": 2,
        "object-curly-spacing": [2, "always"],
        "operator-linebreak": [2, "before", { overrides: { "=": "after" } }],
        "quote-props": [2, "consistent-as-needed"],
        "quotes": [2, "double", { allowTemplateLiterals: true }],
        "semi": [2, "always"],
        "semi-spacing": 2,
        "space-before-blocks": 2,
        "space-in-parens": 2,
        "spaced-comment": 2,
        "yoda": 2
    }
};


src/pretty-money.ts
interface FormatOptions {
    /**
     * The string to be used as currency symbol.
     *
     * It can be the respective sign (like "$"), currency code (like "GBP") or a word (like "peso").
     */
    currency?: string;

    /**
     * Sets the strategy to calculate the amount of decimal places.
     */
    decimalDelimiter?: string;

    /**
     * Sets the amount of decimal places.
     *
     * - "fixed" — the amount of places will always stay at `maxDecimal`. `minDecimal` has no effect.
     * - "fluid" — the amount of places will stay at any number between `minDecimal` and `maxDecimal`, in order not to have trailing zeros.
     * - "minmax" — the amount of places will stay at `maxDecimal` unless it's possible to be at `minDecimal` without having trailing zeros.
     */
    decimals?: "fluid"|"minmax"|"fixed";

    /**
     * The maximum number of decimal places allowed in the number.
     */
    maxDecimal?: number;

    /**
     * The minimum number of decimal places allowed in the number. Has no effect when `decimals` is set to `"fixed"`.
     */
    minDecimal?: number;

    /**
     * Sets the position of the currency symbol with respect to the number.
     */
    position?: "before"|"after";

    /**
     * Sets whether there should be a space between the number and the currency symbol.
     */
    spaced?: boolean;

    /**
     * A string that separates the thousands of the number.
     */
    thousandsDelimiter?: string;
}

const defaultOpts: FormatOptions = {
    currency: "",
    position: "after",
    spaced: true,
    decimals: "minmax",
    minDecimal: 0,
    maxDecimal: 2,
    decimalDelimiter: ".",
    thousandsDelimiter: ""
};

type prettifyFactory = {
    (options: FormatOptions, number: number|string): string;
    (options: FormatOptions): prettify;
    (): prettify;
}

type prettify = {
    (number: number|string): string;
}

/**
 * Returns a curried function to prettify a number according to the given format
 *
 * ## Usage
 * ```
 * > const euros = prettify({currency: "EUR"});
 * > euros("12.345")
 * 12.34 EUR
 *
 * > const rubles = prettify({
 *     currency: "₽",
 *     decimals: "fixed",
 *     decimalDelimiter: ",",
 *     thousandsDelimiter: " "
 * });
 * > rubles(56789)
 * 56 789,00 ₽
 * ```
 *
 * @param options - formatting options
 * @returns the formatting function
 */
function prettyMoney(options: FormatOptions): (n: number|string) => string;

/**
 * Prettifies a number according to the given format
 *
 * ## Usage
 * ```
 * > prettify({currency: "USD"}, 5);
 * 5 USD
 *
 * > prettify({
 *     currency: "₽",
 *     decimals: "fixed",
 *     decimalDelimiter: ",",
 *     thousandsDelimiter: " "
 * }, "56789.0");
 * 56 789,00 ₽
 * ```
 *
 * @param options - formatting options
 * @param number - the number to be currency-formatted
 * @returns the format results
 */
function prettyMoney(options: FormatOptions, number: number|string): string;


function prettyMoney(options: FormatOptions, number?: number | string): string | ((n: number|string) => string) {
    const _opts: FormatOptions = {
        ...defaultOpts,
        ...options
    };

    function prettify(number: number | string): string {
        number = Number(number);

        if (isNaN(number)) {
            console.warn("The number input could not be interpreted");
            number = "NaN";
        } else {
            const tens = Math.pow(10, _opts.maxDecimal);
            number = Math.floor(number * tens).toString();
            const splitIdx = number.length - _opts.maxDecimal;
            const wholePart = number.slice(0, splitIdx).replace(/(\d)(?=(\d{3})+(?!\d))/g, "$1" + _opts.thousandsDelimiter);
            let decimalPart = number.slice(splitIdx);

            if (_opts.decimals === "fluid" || (_opts.decimals === "minmax" && decimalPart.slice(_opts.minDecimal).match(/^0*$/))) {
                decimalPart = decimalPart.slice(0, _opts.minDecimal) + decimalPart.slice(_opts.minDecimal).replace(/0*$/, "");
            }

            number = wholePart + (decimalPart === "" ? "" : _opts.decimalDelimiter) + decimalPart;
        }

        return (_opts.position === "before" ? _opts.currency : "")
            + (_opts.position === "before" && _opts.spaced && _opts.currency !== "" ? " " : "")
            + number
            + (_opts.position === "after" && _opts.spaced && _opts.currency !== "" ? " " : "")
            + (_opts.position === "after" ? _opts.currency : "");
    }

    if (number === undefined) {
        return prettify;
    }

    return prettify(number);
}

export default prettyMoney;


src/pretty-money.js
/**
 * @typedef {Object} FormatOptions
 * 
 * @property {string} [currency=""] The string to be used as currency symbol.
 *
 * It can be the respective sign (like "$"), currency code (like "GBP"),
 * or a word (like "peso").
 * 
 * @property {string} [decimalDelimiter="."] The string that separates the
 * integer and the fractional parts of the number.
 * 
 * @property {"fluid"|"minmax"|"fixed"} [decimals="minmax"] Sets the amount of
 * decimal places the number will have.
 *
 * - "fixed" — the amount of places will always stay at `maxDecimal`.
 *   `minDecimal` has no effect.
 * - "fluid" — the amount of places will stay at any number between `minDecimal`
 *   and `maxDecimal`, in order not to have trailing zeros.
 * - "minmax" — the amount of places will stay at `maxDecimal` unless it's
 *   possible to be at `minDecimal` without having trailing zeros.
 * 
 * @property {number} [maxDecimal=2] The maximum number of decimal places
 * allowed in the number.
 * 
 * @property {number} [minDecimal=2] The minimum number of decimal places
 * allowed in the number. Has no effect when `decimals` is set to `"fixed"`.
 * 
 * @property {"before"|"after"} [position="after"] Sets the position of the
 * currency symbol with respect to the number.
 * 
 * @property {boolean} [spaced=true] Sets whether there should be a space
 * between the number and the currency symbol.
 * 
 * @property {string} [thousandsDelimiter=""] A string that separates the
 * thousands of the number.
 */

/**
 * Default options for formatting.
 * 
 * @type {FormatOptions}
 */
const defaultOpts = {
	currency: "",
	position: "after",
	spaced: true,
	decimals: "minmax",
	minDecimal: 0,
	maxDecimal: 2,
	decimalDelimiter: ".",
	thousandsDelimiter: ""
};

/**
 * Prettifies a number according to the previously defined format.
 * 
 * @callback PrettifyFunction
 * @param {number|string} number the number to be currency formatted
 * @returns {string} formatted number
 */

/**
 * Returns a curried function to prettify a number according to the given format
 *
 * ## Usage
 * ```
 * > const euros = prettify({currency: "EUR"});
 * > euros("12.345")
 * "12.34 EUR"
 *
 * > const rubles = prettify({
 *     currency: "₽",
 *     decimals: "fixed",
 *     decimalDelimiter: ",",
 *     thousandsDelimiter: " "
 *   });
 * > rubles(56789)
 * "56 789,00 ₽"
 * ```
 * @callback PrettifyFunctionFactory
 * @param {FormatOptions} options formatting options
 * @returns {PrettifyFunction} the formatting function
 */

/**
 * Prettifies a number according to the given format
 *
 * ## Usage
 * ```
 * > prettify({currency: "USD"}, 5);
 * "5 USD"
 *
 * > prettify({
 *     currency: "₽",
 *     decimals: "fixed",
 *     decimalDelimiter: ",",
 *     thousandsDelimiter: " "
 *   }, "56789.0");
 * "56 789,00 ₽"
 * ```
 * @callback PrettyMoneyFunction
 * @param {FormatOptions} options formatting options
 * @param {number|string} number the number to be currency formatted
 * @returns {string} the format results
 */

/**
 * @type {PrettifyFunctionFactory & PrettyMoneyFunction} 
 */
const prettyMoney = function (options, number) {
	/** @type {FormatOptions} */
	const _opts = {
		...defaultOpts,
		...options
	};

	function prettify(number) {
		number = Number(number);

		if (isNaN(number)) {
			number = "NaN";
		} else {
			const tens = Math.pow(10, _opts.maxDecimal);
			number = Math.floor(number * tens).toString();

			// Code taken from lodash.repeat (MIT license)
			// Source: https://github.com/lodash/lodash/blob/2f79053d7bc7c9c9561a30dda202b3dcd2b72b90/repeat.js#L20
			// In particular, the exponentiation by squaring algorithm was taken
			let zeroes = _opts.maxDecimal + 1 - number.length;
			if (zeroes > 0){
				let padWith = "0";
				do {
					if (zeroes % 2 != 0) {
						number = padWith + number;
					}
					zeroes = Math.floor(zeroes / 2);
					if (zeroes != 0) {
						padWith += padWith;
					}
				} while (zeroes != 0);
			}

			const splitIdx = number.length - _opts.maxDecimal;
			const wholePart = number.slice(0, splitIdx).replace(/(\d)(?=(\d{3})+(?!\d))/g, "$1" + _opts.thousandsDelimiter);
			let decimalPart = number.slice(splitIdx);

			if (_opts.decimals === "fluid" || (_opts.decimals === "minmax" && decimalPart.slice(_opts.minDecimal).match(/^0*$/))) {
				decimalPart = decimalPart.slice(0, _opts.minDecimal) + decimalPart.slice(_opts.minDecimal).replace(/0*$/, "");
			}

			number = wholePart + (decimalPart === "" ? "" : _opts.decimalDelimiter) + decimalPart;
		}

		return (_opts.position === "before" ? _opts.currency : "")
						+ (_opts.position === "before" && _opts.spaced && _opts.currency !== "" ? " " : "")
						+ number
						+ (_opts.position === "after" && _opts.spaced && _opts.currency !== "" ? " " : "")
						+ (_opts.position === "after" ? _opts.currency : "");
	}

	if (number === undefined) {
		return prettify;
	}

	return prettify(number);
};

export default prettyMoney;
